### ISSUE TITLE
Wrong binary conversion on x86 (32bit intel) arch causing test failures

### ISSUE THREAD

**Issue by jackorp** [ged/ruby-pg#545]
Trying to package the newest (v1.5.4) ruby-pg gem for Fedora, I am hitting a test failure on 32bit intel arch (x86) related to conversion to binary and back with the `Time` class.

The failing test is `Basic type mapping PG::BasicTypeMapBasedOnResult with usage of result oids for copy encoder selection can type cast #copy_data input with encoder to format 1` here: https://github.com/ged/ruby-pg/blob/v1.5.4/spec/pg/basic_type_map_based_on_result_spec.rb#L121

This happens only with `format 1` which is using the `PG::BinaryEncoder`

The failure:
```
  1) Basic type mapping PG::BasicTypeMapBasedOnResult with usage of result oids for copy encoder selection can type cast #copy_data input with encoder to format 1
     Failure/Error:
       expect( res.values ).to eq( [
       	["\\xff000a0d27", "123", "2023-03-17 03:04:05.678912", "1990-12-17 15:14:45", "12.345", "-1.2345e+168", "2055-12-31", "1234-08-31"],
       	["\\x202078797a2020", "-444", "infinity", "-infinity", "-Infinity", "NaN", "infinity", "-infinity"]
       ] )

       expected: [["\\xff000a0d27", "123", "2023-03-17 03:04:05.678912", "1990-12-17 15:14:45", "12.345", "-1.2345e+168", "2055-12-31", "1234-08-31"], ["\\x202078797a2020", "-444", "infinity", "-infinity", "-Infinity", "NaN", "infinity", "-infinity"]]
            got: [["\\xff000a0d27", "123", "1999-12-31 23:31:17.070656", "1999-12-31 23:50:53.061952", "12.345", "-1.2345e+168", "2055-12-31", "1234-08-31"], ["\\x202078797a2020", "-444", "infinity", "-infinity", "-Infinity", "NaN", "infinity", "-infinity"]

       (compared using ==)
     # ./spec/pg/basic_type_map_based_on_result_spec.rb:137:in `block (5 levels) in <top (required)>'
     # /builddir/build/BUILD/spec/helpers.rb:56:in `block in included'
```

We can see the weird dates are starting with `1999-12-31`, but more importantly only for data created via the `Time` class, notice that the `Date` class does not suffer from this.

Env info:
```
PGSQL version: 15.4
Fedora: Rawhide (F40)
ruby -v: ruby 3.2.2 (2023-03-30 revision e51014f9c0) [i386-linux]
gem pg version:  1.5.4
```

**Comment by jackorp** [comment]
I was able to distill it down without using any Database connection.

```Ruby
require 'pg'

enc_ts = PG::BinaryEncoder::Timestamp.new().encode(Time.utc(2023,1,1,1,1, 1.22221))

p PG::BinaryDecoder::Timestamp.new().decode(enc_ts)
# => 2000-01-01 00:09:48.19821 UTC
```

From what I observed poking around with prints in C, the int64_t value of "timestamp" that encoder spits out gets into decoder unchanged, but not sure if it is the correct value in the first place.

I also noticed that the value I get from encoder differs between 64bit and 32bit.

64bit:
```ruby
require 'pg'
p PG::BinaryEncoder::Timestamp.new().encode(Time.utc(2023,1,1,1,1, 1.22221))
# => "\x00\x02\x94(#\x0F1B"
```

32bit:
```ruby
require 'pg'
p PG::BinaryEncoder::Timestamp.new().encode(Time.utc(2023,1,1,1,1, 1.22221))
# => "\x00\x00\x00\x00#\x0F1B"
```

**Comment by jackorp** [comment]
Actually there is another related failure:
```
  2) PG::Type derivations PG::SimpleCoder#encode binary timestamps encodes timestamps as local time
     Failure/Error:
       expect( binaryenc_timestamplocal.encode(Time.new(2000,1,1)) ).
       	to eq( "\x00" * 8 )

       expected: "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
            got: "\x00\x00\x00\x01\x00\x00\x00\x00"

       (compared using ==)
     # ./spec/pg/type_spec.rb:416:in `block (5 levels) in <top (required)>'
     # /builddir/build/BUILD/spec/helpers.rb:56:in `block in included'
```

**Comment by larskanis** [comment]
Thanks @jackorp for reporting this and for the fix in #547. I already made a fix in #548 yesterday, but I think your fix is better. While testing the issue I found another issue on 32 bit , which is fixed by the second commit in #548.

**Comment by jackorp** [comment]
Thanks for the swift work! Feel free to take my commit into your PR if you'd like.

I personally like the retyping as close to the variable as possible, it feels more robust and less prone to being compiler specific or less prone to hitting a potential compiler bug, but I only have my feelings and not empirical data to back it up :)

Thanks again for your work on this and the other issue you found :+1:
